esphome/components/bl0940/bl0940.h

```diff
@@ -1,58 +1,37 @@
 #pragma once
 
 #include "esphome/core/component.h"
-#include "esphome/components/uart/uart.h"
+#include "esphome/core/datatypes.h"
+#include "esphome/core/defines.h"
+#ifdef USE_BUTTON
+#include "esphome/components/button/button.h"
+#endif
+#ifdef USE_NUMBER
+#include "esphome/components/number/number.h"
+#endif
 #include "esphome/components/sensor/sensor.h"
+#include "esphome/components/uart/uart.h"
 
 namespace esphome {
 namespace bl0940 {
 
-static const float BL0940_PREF = 1430;
-static const float BL0940_UREF = 33000;
-static const float BL0940_IREF = 275000;  // 2750 from tasmota. Seems to generate values 100 times too high
-
-// Measured to 297J  per click according to power consumption of 5 minutes
-// Converted to kWh (3.6MJ per kwH). Used to be 256 * 1638.4
-static const float BL0940_EREF = 3.6e6 / 297;
-
-struct ube24_t {  // NOLINT(readability-identifier-naming,altera-struct-pack-align)
-  uint8_t l;
-  uint8_t m;
-  uint8_t h;
-} __attribute__((packed));
-
-struct ube16_t {  // NOLINT(readability-identifier-naming,altera-struct-pack-align)
-  uint8_t l;
-  uint8_t h;
-} __attribute__((packed));
-
-struct sbe24_t {  // NOLINT(readability-identifier-naming,altera-struct-pack-align)
-  uint8_t l;
-  uint8_t m;
-  int8_t h;
-} __attribute__((packed));
-
 // Caveat: All these values are big endian (low - middle - high)
-
-union DataPacket {  // NOLINT(altera-struct-pack-align)
-  uint8_t raw[35];
-  struct {
-    uint8_t frame_header;  // value of 0x58 according to docs. 0x55 according to Tasmota real world tests. Reality wins.
-    ube24_t i_fast_rms;    // 0x00
-    ube24_t i_rms;         // 0x04
-    ube24_t RESERVED0;     // reserved
-    ube24_t v_rms;         // 0x06
-    ube24_t RESERVED1;     // reserved
-    sbe24_t watt;          // 0x08
-    ube24_t RESERVED2;     // reserved
-    ube24_t cf_cnt;        // 0x0A
-    ube24_t RESERVED3;     // reserved
-    ube16_t tps1;          // 0x0c
-    uint8_t RESERVED4;     // value of 0x00
-    ube16_t tps2;          // 0x0c
-    uint8_t RESERVED5;     // value of 0x00
-    uint8_t checksum;      // checksum
-  };
+struct DataPacket {
+  uint8_t frame_header;    // value of 0x58 according to en docs but 0x55 in cn docs and Tasmota real world tests
+  uint24_le_t i_fast_rms;  // 0x00
+  uint24_le_t i_rms;       // 0x04
+  uint24_t RESERVED0;      // reserved
+  uint24_le_t v_rms;       // 0x06
+  uint24_t RESERVED1;      // reserved
+  int24_le_t watt;         // 0x08
+  uint24_t RESERVED2;      // reserved
+  uint24_le_t cf_cnt;      // 0x0A
+  uint24_t RESERVED3;      // reserved
+  uint16_le_t tps1;        // 0x0c
+  uint8_t RESERVED4;       // value of 0x00
+  uint16_le_t tps2;        // 0x0c
+  uint8_t RESERVED5;       // value of 0x00
+  uint8_t checksum;        // checksum
```

Comment:
I've got no experience using `uint24` types, but, in general, you'll likely see slightly lower memory usage by arranging from widest to narrowest:
```suggestion
  uint24_le_t i_fast_rms;  // 0x00
  uint24_le_t i_rms;       // 0x04
  uint24_t RESERVED0;      // reserved
  uint24_le_t v_rms;       // 0x06
  uint24_t RESERVED1;      // reserved
  int24_le_t watt;         // 0x08
  uint24_t RESERVED2;      // reserved
  uint24_le_t cf_cnt;      // 0x0A
  uint24_t RESERVED3;      // reserved
  uint16_le_t tps1;        // 0x0c
  uint16_le_t tps2;        // 0x0c
  uint8_t RESERVED4;       // value of 0x00
  uint8_t RESERVED5;       // value of 0x00
  uint8_t checksum;        // checksum
  uint8_t frame_header;    // value of 0x58 according to en docs but 0x55 in cn docs and Tasmota real world tests
```

Reply:
don't think this can be changed as it is a representation of the datawacket received from the sensor